server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo9_0_0.java

```diff
@@ -83,4 +89,71 @@ protected void addAccountConsoleClient(RealmModel realm) {
             client.addProtocolMapper(audienceMapper);
         }
     }
+
+
+    protected void migrateAuthenticationFlowsWithAlternativeRequirements(RealmModel realm) {
+        for (AuthenticationFlowModel flow : realm.getAuthenticationFlows()) {
+
+            // Suffix used just to avoid name conflicts
+            int suffix = 0;
+            boolean requiredFound = false;
+            LinkedList<AuthenticationExecutionModel> alternativesAfterRequired = new LinkedList<>();
+            for (AuthenticationExecutionModel execution : realm.getAuthenticationExecutions(flow.getId())) {
+                if (AuthenticationExecutionModel.Requirement.REQUIRED.equals(execution.getRequirement())) {
```

Comment:
For a flow like...
- 1) ExecutionA, conditional
- 2) ExecutionB, alternative
- 3) ExecutionC, alternative
...there would be no migration. But effectively, this flow is now ignoring step 2 and 3 as no migration is applied (due to org.keycloak.authentication.DefaultAuthenticationFlow.processFlow():
`            else if (execution.isRequired() || execution.isConditional()) {
                requiredList.add(execution);
            }
`

Reply:
You're right. Will improve that.